### PR TITLE
native: add committee change events

### LINKED ISSUE #3326: Port CommitteeChanged events
## Is your feature request related to a problem? Please describe.
I'm always frustrated when we're lagging behind the C# node in protocol development.

## Describe the solution you'd like
Port https://github.com/neo-project/neo/pull/3158.

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Port neo-project/neo#3158.

Close #3326

### CODE REVIEW

File: pkg/core/native/native_neo.go [hunk 1]
```diff
@@ -425,11 +429,54 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
+		for i, member := range cache.committee {
+			pub, err := member.PublicKey()
+			if err != nil {
+				return fmt.Errorf("failed to get public key: %w", err)
+			}
+			newCommitteePublicKeys[i] = pub
+		}
+		// Check if the committee has changed by comparing public keys directly
+		committeeChanged := len(prevCommitteePublicKeys) != len(newCommitteePublicKeys)
+		for i := 0; i < len(prevCommitteePublicKeys) && !committeeChanged; i++ {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:459]: It's the reason why your test is failing.

File: pkg/core/native/native_test/neo_test.go [hunk 2]
```diff
@@ -117,6 +117,60 @@ func TestNEO_CandidateEvents(t *testing.T) {
...
+	}
+	//for _, tx := range txes {
+	//	e.CheckHalt(t, tx.Hash(), stackitem.Make(true))
+	//}
+
+	prevCommitteePublicKey, err := keys.NewPublicKeysFromStrings(cfg.StandbyCommittee)
+	require.NoError(t, err)
+	fmt.Println(prevCommitteePublicKey)
+
+	advanceChain(t)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:159]: Manually advance chain up to the block whose index is divisible by committee size. Then check that block's OnPersist execution result contains expected event (change from StandBy committee to the new one). Use `e.Chain.GetAppExecResults` with *block* hash and `trigger.OnPersist` to get block's AER.

File: pkg/core/native/native_test/neo_test.go [hunk 3]
```diff
@@ -117,6 +117,60 @@ func TestNEO_CandidateEvents(t *testing.T) {
...
+	newCommitteePublicKeys := make([]*keys.PublicKey, committeeSize)
+	for i, member := range neoCommitteeInvoker.Signers {
+		newCommitteePublicKeys[i] = member.(neotest.SingleSigner).Account().PublicKey()
+	}
+	tx := neoCommitteeInvoker.WithSigners(candidates[0]).Invoke(t, true, "unregisterCandidate", candidates[0].(neotest.SingleSigner).Account().PublicKey().Bytes())
+	e.CheckTxNotificationEvent(t, tx, 0, state.NotificationEvent{
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:165]: It's wrong API, CommitteeChange event is emitted during edge block OnPersist script execution. Ref. comment above for right API.

File: pkg/core/native/native_neo.go [hunk 4]
```diff
@@ -425,17 +429,52 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
 
+// checkCommitteeChange checks if the committee has changed and returns the previous and new committee as stackitem.Item.
+func (n *NEO) checkCommitteeChange(cache *NeoCache) (bool, stackitem.Item, stackitem.Item) {
+	if len(cache.committee) != len(cache.newEpochCommittee) {
+		return true, cache.committee.toNotificationItem(), cache.newEpochCommittee.toNotificationItem()
+	}
+
+	// Compare every key in the new committee to every key in the previous committee.
+	for i := 0; i < len(cache.newEpochCommittee); i++ {
+		keyFound := false
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:460]: Don't need `keyFound` variable. Return immediately if the difference is found.

File: pkg/core/native/native_neo.go [hunk 5]
```diff
@@ -425,17 +429,52 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
+			}))
+		}
 		// We need to put in storage anyway, as it affects dumps
 		ic.DAO.PutStorageItem(n.ID, prefixCommittee, cache.committee.Bytes(ic.DAO.GetItemCtx()))
 	}
 	return nil
 }
 
+// checkCommitteeChange checks if the committee has changed and returns the previous and new committee as stackitem.Item.
+func (n *NEO) checkCommitteeChange(cache *NeoCache) (bool, stackitem.Item, stackitem.Item) {
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:453]: Don't need a separate function, because no one else uses it and it's small. Besides, its result strictly depends on committee cache update and may be wrong if called in wrong place.

File: pkg/core/native/native_neo.go [hunk 6]
```diff
@@ -425,17 +429,52 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
+	if len(cache.committee) != len(cache.newEpochCommittee) {
+		return true, cache.committee.toNotificationItem(), cache.newEpochCommittee.toNotificationItem()
+	}
+
+	// Compare every key in the new committee to every key in the previous committee.
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:458]: Comments are not needed in this function, this code is clear and well-understandable.

File: pkg/core/native/native_test/neo_test.go [hunk 7]
```diff
@@ -117,6 +117,54 @@ func TestNEO_CandidateEvents(t *testing.T) {
...
+
+	// Advance the chain to trigger committee recalculation and potential change
+	for (block.Index)%uint32(committeeSize) != 0 {
+		block = neoCommitteeInvoker.AddNewBlock(t)
+	}
+	// Check for CommitteeChanged event in the last persisted block's AER
+	blockHash := e.Chain.CurrentBlockHash()
+	aer, err := e.Chain.GetAppExecResults(blockHash, trigger.OnPersist)
+	require.NoError(t, err)
+	require.NotEmpty(t, aer)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:158]:
```suggestion
	require.Equal(t, 1, len(aer))
```

File: pkg/core/native/native_test/neo_test.go [hunk 8]
```diff
@@ -117,6 +117,54 @@ func TestNEO_CandidateEvents(t *testing.T) {
...
+	found := false
+	for _, event := range aer[0].Events {
+		if event.Name == "CommitteeChanged" {
+			found = true
+			break
+		}
+	}
+	require.True(t, found, "CommitteeChanged event not found")
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:166]: It's not enough. The block is known and empty, the order of events is strictly fixed. Thus, instead of this check we need to ensure that the CommitteeChanged event is placed at the correct index and contains exactly the expected stackitems. Old committee is chain's StandBy committee, those public keys may be retrieved from the chain's config, and the new committee should be constructed from the list of registered users.

File: pkg/core/native/native_neo.go (hunk 4 above)
Comment (AliceInHunterland) [reply pkg/core/native/native_neo.go]: cache.newEpochCommittee is sorted, right?

File: pkg/core/native/native_neo.go [hunk 9]
```diff
@@ -425,6 +429,16 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
+
+		for i := 0; i < len(cache.newEpochCommittee); i++ {
+			if cache.newEpochCommittee[i].Key != cache.committee[i].Key || len(cache.newEpochCommittee) != len(cache.committee) {
+				ic.AddNotification(n.Hash, "CommitteeChanged", stackitem.NewArray([]stackitem.Item{
+					cache.committee.toNotificationItem(), cache.newEpochCommittee.toNotificationItem(),
+				}))
+				break
+			}
+		}
+
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:441]:
```suggestion
		// Cached newEpoch* values always have proper value set (either by PostPersist
		// during the last epoch block handling or by initialization code).
		var oldCommittee, newCommittee stackitem.Item
		for i := 0; i < len(cache.committee); i++ {
			if cache.newEpochCommittee[i].Key != cache.committee[i].Key ||
				(i == 0 && len(cache.newEpochCommittee) != len(cache.committee)) {
				oldCommittee, newCommittee = cache.committee.toNotificationItem(), cache.newEpochCommittee.toNotificationItem()
				break
			}
		}

		cache.nextValidators = cache.newEpochNextValidators
		cache.committee = cache.newEpochCommittee
		cache.committeeHash = cache.newEpochCommitteeHash
		cache.votesChanged = false

		// We need to put in storage anyway, as it affects dumps
		ic.DAO.PutStorageItem(n.ID, prefixCommittee, cache.committee.Bytes(ic.DAO.GetItemCtx()))

		if oldCommittee != nil {
			ic.AddNotification(n.Hash, "CommitteeChanged", stackitem.NewArray([]stackitem.Item{
				oldCommittee, newCommittee,
			}))
		}
```

Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:441]: We'd better place notification in the end in order not to break the compatibility if we're to add more code to OnPersist in future.

File: pkg/core/native/native_test/neo_test.go [hunk 10]
```diff
@@ -118,6 +118,70 @@ func TestNEO_CandidateEvents(t *testing.T) {
...
+
+	expectedOldCommitteePublicKeys, err := keys.NewPublicKeysFromStrings(cfg.StandbyCommittee)
+	require.NoError(t, err)
+	expectedOldCommitteeStackItems := make([]stackitem.Item, len(expectedOldCommitteePublicKeys))
+	for i, pubKey := range expectedOldCommitteePublicKeys {
+		expectedOldCommitteeStackItems[i] = stackitem.NewByteArray(pubKey.Bytes())
+	}
+	oldCommitteeStackItem := aer[0].Events[0].Item.Value().([]stackitem.Item)[0].(*stackitem.Array)
+	for i, item := range oldCommitteeStackItem.Value().([]stackitem.Item) {
+		require.Equal(t, expectedOldCommitteeStackItems[i].(*stackitem.ByteArray).Value().([]byte), item.Value().([]byte))
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:173]: It's better to use `assert` here.

File: pkg/services/rpcsrv/server_test.go [hunk 11]
```diff
@@ -87,7 +87,7 @@ const (
 	faultedTxHashLE                   = "82279bfe9bada282ca0f8cb8e0bb124b921af36f00c69a518320322c6f4fef60"
 	faultedTxBlock             uint32 = 23
 	invokescriptContractAVM           = "VwIADBQBDAMOBQYMDQIODw0DDgcJAAAAAErZMCQE2zBwaEH4J+yMqiYEEUAMFA0PAwIJAAIBAwcDBAUCAQAOBgwJStkwJATbMHFpQfgn7IyqJgQSQBNA"
-	block20StateRootLE                = "858c873539d6d24a70f2be13f9dafc61aef2b63c2aa16bb440676de6e44e3cf1"
+	block20StateRootLE                = "397c69adbc0201d59623fa913bfff4a2da25c792c484d1d278c061709f2c21cf"
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/server_test.go:90]: Not related to Neo changes, should be done in a separate commit.

**Review by AnnaShaleva — COMMENTED** [review]
Otherwise LGTM.

File: pkg/core/native/native_test/neo_test.go (hunk 10 above)
Comment (AnnaShaleva) [reply pkg/core/native/native_test/neo_test.go:173]: And below.

File: pkg/core/native/native_neo.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go]: Right, don't need to re-sort.
